### ISSUE TITLE
possible memory leak when using chart in a data template

### ISSUE THREAD

**Issue by d4art** [beto-rodriguez/Live-Charts#457]
#### How to reproduce?

I just made a quick listview bound to an observablecollection with some data. The listview has a datatemplate just containing an empty cartesianchart. After removing and adding data to the collection the memory leak occurs as the old charts are not garbage collected. 

View

```
<Window x:Class="LiveChartsTest.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:LiveChartsTest"
        xmlns:charts="clr-namespace:LiveCharts.Wpf;assembly=LiveCharts.Wpf"
        mc:Ignorable="d"
        Title="MainWindow" Height="350" Width="525">
    <Grid>
        <Grid.DataContext>
            <local:MainMindowViewModel/>
        </Grid.DataContext>
        <ListView ItemsSource="{Binding List}">
            <ListView.ItemTemplate>
                <DataTemplate>
                    <charts:CartesianChart Width="100" Height="100"/>
                </DataTemplate>
            </ListView.ItemTemplate>
        </ListView>
    </Grid>
</Window>
```
ViewModel

```
namespace LiveChartsTest {
    public class MainMindowViewModel {

        public ObservableCollection<string> List {
            get; set;
        }

        private Random r = new Random();

        public MainMindowViewModel() {
            List = new ObservableCollection<string>();
            List.Add("a");
            List.Add("b");
            List.Add("c");
            List.Add("d");
            List.Add("e");

            new Thread(() => {
                while (true) {
                    Thread.Sleep(200);
                    Application.Current.Dispatcher.Invoke((Action)(() => {
                        List.RemoveAt(0);
                    }));

                    Thread.Sleep(200);
                    Application.Current.Dispatcher.Invoke((Action)(() => {
                        List.Add("a"+(r.NextDouble()));
                    }));


                }
            }).Start();
        }
    }
}
```

#### Extra notes

I think the problem is located in ChartUpdater

```
namespace LiveCharts.Wpf.Components
{
    internal class ChartUpdater : LiveCharts.ChartUpdater
    {
        public ChartUpdater(TimeSpan frequency)
        {
            Timer = new DispatcherTimer {Interval = frequency};

            Timer.Tick += (sender, args) => {
                UpdaterTick(RequiresRestart);
            };
        }
```

UpdaterTick is never removed and or stopped. For a quick fix (I dont know if I broke something else), I just added a Stop Method to ChartUpdater and modified the Run Method.

 ```
public override void Run(bool restartView = false, bool updateNow = false)
        {
            if (Timer == null) return;
            if (updateNow || Chart.View.IsMocked)
            {
                UpdaterTick(restartView);
                return;
            }

            RequiresRestart = restartView || RequiresRestart;
            if (IsUpdating) return;

            IsUpdating = true;
            Timer.Start();
        }

        public void Stop() {
            Timer.Tick -= (sender, args) => {
                UpdaterTick(RequiresRestart);
            };
            Timer = null;
        }
```

I called this Method in the Wpf.Charts.Base.Chart class, which looks a bit ugly..

```
 private void OnIsVisibleChanged(object sender, DependencyPropertyChangedEventArgs dependencyPropertyChangedEventArgs)
        {
            if (!(bool)dependencyPropertyChangedEventArgs.NewValue) {
                ((LiveCharts.Wpf.Components.ChartUpdater)this.ChartCoreModel.Updater).Stop();
            }
            Model.ControlSize = new CoreSize(ActualWidth, ActualHeight);

            Model.Updater.Run();
            PrepareScrolBar();
        }
```
After this fix, the leak is gone. Maybe the charts are not supposed to be used in datatemplates or I need to take a different approach. As a background information I had a list with 10 charts and I implemented drag and drop for this list.

#### I am using

<table>
<tr>
    <td>Live-Charts version</td>
    <td>0.9.2</td>
</tr>
<tr>
    <td>.Net Version</td>
    <td>4.6.1</td>
</tr>
<tr>
    <td>Windows</td>
    <td>10</td>
</tr>
</table>

**Comment by d4art** [comment]
I just saw that a modification in Wpf.Charts.Base.Chart
```
 private void OnIsVisibleChanged(object sender, DependencyPropertyChangedEventArgs dependencyPropertyChangedEventArgs)
        {
            if (!(bool)dependencyPropertyChangedEventArgs.NewValue) {
                IsMocked = true;
            }
            Model.ControlSize = new CoreSize(ActualWidth, ActualHeight);

            Model.Updater.Run();
            PrepareScrolBar();
        }
```
seems to be enough

**Comment by beto-rodriguez** [comment]
@d4art thank you very much, you were right the memory leak was related to the updater `Timer.Tick` event

to prevent braking something else, I remover the handler when the control is unloaded instead, this fix will be included in the next version  of the library (0.9.4)

on the constuctor of  Wpf. Base.Chart:

    Unloaded += (sender, args) =>
             {
                 //remove updater timer from memory
                 var updater = (Components.ChartUpdater) Model.Updater;
                 updater.Timer.Tick -= updater.OnTimerOnTick;
                 updater.Timer.Stop();
                 updater.Timer.IsEnabled = false;
                 updater.Timer = null;
             };

thanks again!
